app/controllers/concerns/dynamic_field_names_cacher.rb

```diff
@@ -0,0 +1,21 @@
+module DynamicFieldTitlesCacher
+  CACHE_KEY = "%{identifier}/%{record_type}/column-keys/%{tail}".freeze
+
+  # Takes an ActiveRecord::Relation so we can make one more query off it
+  def collection_field_names(collection, identifier = "projects-#{@current_project.id}")
```

Comment:
I don't know about working so hard as part of the argument? what about `identifier=nil` and then later `identifier ||= "projects-" to keep the signature a bit cleaner?